mkosi

```diff
@@ -268,59 +271,134 @@ def prepare_esp(args, loopdev):
+        subprocess.run(["cryptsetup", "open", "--type", "luks", dev, luks_name], input=luks_key, check=True)
+    elif luks_key['type'] == 'file':
+        subprocess.run(["cryptsetup", "--key-file", luks_key["content"], "open", "--type", "luks", dev, luks_name], check=True)
+
+    return os.path.join("/dev/mapper", luks_name)
+
+@contextlib.contextmanager
+def luks_try_format(loopdev_part, part_name, luks_keys):
+    luks_key = luks_keys.get(part_name, None)
+    if luks_key:
+        print_step("Setting up LUKS encryption on {} partition...".format(part_name))
+        luks_format(loopdev_part, luks_key)
+        dev = luks_open(loopdev_part, luks_key)
+    else:
+        dev = loopdev_part
+
+    try:
+        yield dev
+    finally:
+        if luks_key:
+            print_step("Closing {} LUKS partition.".format(part_name))
+            luks_close(dev)
+
+def luks_close(dev):
+    subprocess.run(["cryptsetup", "luksClose", dev], check=True)
+
+def prepare_root(args, loopdev, luks_keys):
     if loopdev is None:
         return
     if args.root_partno is None:
         return
     if args.output_format == OutputFormat.raw_squashfs:
         return
 
-    print_step("Formatting root partition...");
-
     if args.output_format == OutputFormat.raw_btrfs:
+        print_step("Formatting root partition...");
         subprocess.run(["mkfs.btrfs", "-Lroot", partition(loopdev, args.root_partno)], check=True)
     else:
-        mkfs_ext4("root", "/", loopdev, args.root_partno)
+        loopdev_part = partition(loopdev, args.root_partno)
+        with luks_try_format(loopdev_part, "root", luks_keys) as dev:
+            print_step("Formatting root partition...");
+            mkfs_ext4("root", "/", dev)
+            print_step("Formatting root partition completed.");
 
-    print_step("Formatting root partition completed.");
-
-def prepare_home(args, loopdev):
+def prepare_home(args, loopdev, luks_keys):
     if loopdev is None:
         return
     if args.home_partno is None:
         return
 
-    print_step("Formatting home partition...");
-
-    mkfs_ext4("home", "/home", loopdev, args.home_partno)
-
-    print_step("Formatting home partition completed.");
+    loopdev_part = partition(loopdev, args.home_partno)
+    with luks_try_format(loopdev_part, "home", luks_keys) as dev:
+        print_step("Formatting home partition...");
+        mkfs_ext4("home", "/home", dev)
+        print_step("Formatting home partition completed.");
 
-def prepare_srv(args, loopdev):
+def prepare_srv(args, loopdev, luks_keys):
     if loopdev is None:
         return
     if args.srv_partno is None:
         return
 
-    print_step("Formatting server data partition...");
+    loopdev_part = partition(loopdev, args.srv_partno)
+    with luks_try_format(loopdev_part, "srv", luks_keys) as dev:
+        print_step("Formatting server data partition...");
+        mkfs_ext4("srv", "/srv", dev)
+        print_step("Formatted server data partition.");
```

Comment:
hmm, blocks of code very similar to this one have been used now multiple times. I figure this should become a new function. mkfs_ext4_with_luks() or so, that wraps all this